src/util/saveArticles.ts

```diff
@@ -236,6 +237,21 @@ async function saveArticle(
   }
 }
 
+export function getArticleUrl(downloader: Downloader, dump: Dump, articleId: string): string {
+  return `${isMainPage(dump, articleId) ? downloader.baseUrlForMainPage : downloader.baseUrl}${encodeURIComponent(articleId)}`
+}
+
+function isMainPage(dump: Dump, articleId: string): boolean {
+  return dump.isMainPage(articleId)
+}
+
+function chooseRenderer(articleId, dump, downloader, desktopRenderer, visualEditorRenderer) {
+  if (isMainPage(dump, articleId) || (downloader.mwCapabilities.veApiAvailable && !downloader.mwCapabilities.desktopRestApiAvailable)) {
```

Comment:
Many things are wrong here:
* Not normal to have this `mwCapabilities` here, this should disappear and rely on `Mediawiki` methods to check
* We should never in the same run have a `desktopRenderer` and a `visualEditorRenderer`, both are Desktop HTML. Remember mainPage should renderer as Desktop when the rest as mobile. You should have one renderer caled "article_renderer" and the other one "mainpage_renderer".

Reply:
To summarize here

The **main page** should use:
**WikimediaDesktop** (api/rest_v1/page/html/) _OR_ **VisualEditor** (api.php?action=visualeditor)

For other pages use:
**WikimediaMobile** (api/rest_v1/page/mobile-html)

There will be two renderers:
1. For the main page (page/html OR visual editor (visual editor is kind of fallback if page/html is not present))
2. For articles (only mobile-htlml when we decide to apply it).

Current implementation:
Renderer shares multiple responsibilities and handles different kinds of conditions to generate article output.

What I want to achieve now to proceed further:
1. Make sure all tests work for HTML (**WikimediaDesktop** or **VisualEditor**)
2. Prepare separate render classes without breaking tests. Currently, there are classes for **WikimediaDesktop** or **VisualEditor**. They have different rendering methods because they manipulate different kinds of data (HTML vs JSON). Also, I still need to figure out how to manage with isMainPage condition, probably in a different PR.

Reply:
Upd:
1. No more `mwCapabilities` from Downloader now. Now it is inside MediaWiki and only the renderer decides which capability to choose.
2. Done.